### ISSUE TITLE
Allow multiple authentication methods per route

### ISSUE THREAD

**Issue by markokocic** [apache/apisix#10196]
### Description

Right now, it is not possible to have multiple authentication methods defined for one route. If you want to expose one service to multiple consumers, where some consumers use basic auth, some key auth and some jwt, you need to define three different routes.

A better approach would be to allow different authentication methods for a route, maybe with some prioritization, e.g. if basic auth params present - use basic, if app-key present, use key auth, if jwt-key present, use jwt-auth, ....

**Comment by madhawa-gunasekara** [comment]
Hi, 
I would like to implement this. 
This feature should be integrated into the core functionality, as all the authentication plugins have “auth” as their plugin type. The implementation can be done by iterating through all the auth plugins, and stopping the iteration once a plugin is accepted and passing it on.

**Comment by shreemaan-abhishek** [comment]
Thanks @madhawa-gunasekara

**Comment by juzhiyuan** [comment]
similar to https://github.com/apache/apisix/issues/6217

**Comment by madhawa-gunasekara** [comment]
Hi @shreemaan-abhishek , 

I have looked into the code, I found two ways to implement this feature.

- Plugin Addition:
Consider integrating an additional plugin, acting as a wrapper, to facilitate multi-authentication seamlessly. This would enhance the existing authentication framework and provide a streamlined solution.

- Modifying the Core Plugin
The second approach involves modifying the [apisix.plugin](https://github.com/apache/apisix/blob/master/apisix/plugin.lua#L1108) to support multiple authentications at the route level. Given our current execution flow, where plugins are prioritized and filtered, this adjustment requires a special case. We will iterate through all plugins associated with a route, identify those related to authentication, and execute them sequentially. This way, we can discern if any authentication method aligns with the incoming request.
